lib/src/common/cache/cache.dart

```diff
@@ -294,6 +372,37 @@ class Cache<TIdentifier, TValue> extends Object with Disposable {
     return new Future.value();
   }
 
+  /// Returns `true` and calls [callback] if there is a cached, unreleased
+  /// [TValue] associated with [id]; returns `false` otherwise.
+  ///
+  /// This does not perform a [get] or [getAsync], and as a result, will not
+  /// affect retention or removal of a [TIdentifier][TValue] pair from the cache.
+  /// If any callbacks are in fight on removal of [TIdentifier] [TValue] pair
+  /// [didRemove] will not event until callback has completed.
+  ///
+  /// If the [Cache] [isOrWillBeDisposed] then a [StateError] is thrown.
+  Future<bool> applyToItem(
+      TIdentifier id, Future<Null> callback(Future<TValue> value)) {
+    _log.finest('applyToItem id: $id');
+    _throwWhenDisposed('applyToItem');
+    if (_isReleased[id] != false) {
+      return new Future.value(false);
+    }
+    final callbackFuture = callback(_cache[id]);
+    if (callbackFuture is Future<Null>) {
+      _applyToItemCallBacks.putIfAbsent(id, () => <Future<Null>>[]);
+      _applyToItemCallBacks[id].add(callbackFuture);
+      callbackFuture.then((_) {
+        _applyToItemCallBacks[id].remove(id);
```

Comment:
Shouldn't this be `.remove(callbackFuture)`?

Might be a good idea to expose a `@visibleForTesting ` version of `_applyToItemCallBacks` to ensure it's getting managed as expected.